src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs

```diff
@@ -13,44 +14,108 @@ namespace Microsoft.CodeAnalysis
+            {
+                char c = hintName[i];
+                if (!UnicodeCharacterUtilities.IsIdentifierPartCharacter(c)
+                    && c != '.'
+                    && c != ','
+                    && c != '-'
+                    && c != '_'
+                    && c != ' '
+                    && c != '('
+                    && c != ')'
+                    && c != '['
+                    && c != ']'
+                    && c != '{'
+                    && c != '}')
+                {
+                    throw new ArgumentException(string.Format(CodeAnalysisResources.HintNameInvalidChar, c, i), nameof(hintName));
+                }
+            }
+
+            hintName = AppendExtensionIfRequired(hintName);
+            lock (_collectionLock)
+            {
+                if (_hintNames.Contains(hintName))
+                {
+                    throw new ArgumentException(CodeAnalysisResources.HintNameUniquePerGenerator, nameof(hintName));
+                }
+
+                _hintNames.Add(hintName);
+                _sourcesAdded.Add(new GeneratedSourceText(hintName, source));
+            }
         }
 
         public void RemoveSource(string hintName)
         {
-            _sourcesAdded.Remove(hintName);
+            hintName = AppendExtensionIfRequired(hintName);
+            lock (_collectionLock)
+            {
+                if (_hintNames.Contains(hintName))
+                {
+                    _hintNames.Remove(hintName);
+                    for (int i = 0; i < _sourcesAdded.Count; i++)
+                    {
+                        // check the hashcodes, as thats the comparison we're using to put the names into _hintNames
+                        if (_sourcesAdded[i].HintName.GetHashCode() == hintName.GetHashCode())
+                        {
+                            _sourcesAdded.Remove(_sourcesAdded[i]);
+                            return;
+                        }
+                    }
+                }
+            }
         }
 
-        public bool Contains(string hintName) => _sourcesAdded.ContainsKey(hintName);
+        public bool Contains(string hintName)
+        {
+            lock (_collectionLock)
+            {
+                return _hintNames.Contains(AppendExtensionIfRequired(hintName));
+            }
+        }
 
         internal ImmutableArray<GeneratedSourceText> ToImmutableAndFree()
         {
-            // https://github.com/dotnet/roslyn/issues/42627: This needs to be consistently ordered
-            ArrayBuilder<GeneratedSourceText> builder = ArrayBuilder<GeneratedSourceText>.GetInstance();
-            foreach (var (hintName, sourceText) in _sourcesAdded)
+            _hintNames.Free();
+            return _sourcesAdded.ToImmutableAndFree();
+        }
+
+        private static string AppendExtensionIfRequired(string hintName)
+        {
+            if (!Path.GetExtension(hintName).Equals(".cs", StringComparison.OrdinalIgnoreCase))
```

Comment:
The code for comparing hint names in general is ordinal equality but the path is ignore case. Should be consistent here.

Reply:
Good catch. I've made all the hint path comparison be case insensitive as we'll need them to be to support windows anyway. Added more test case theory data to cover it.

---
In reply to: [435565786](https://github.com/dotnet/roslyn/pull/44768#discussion_r435565786) [](ancestors = 435565786)